core/src/test/java/kafka/test/api/ShareConsumerTest.java

```diff
@@ -1069,100 +1055,50 @@ public void testMultipleConsumersInMultipleGroupsConcurrentConsumption(String pe
-            for (CompletableFuture<Integer> future : futures1) {
-                totalResult1 += future.get();
-            }
+            consumeMessagesFutures1.add(CompletableFuture.supplyAsync(() ->
+                    consumeMessages(totalMessagesConsumedGroup1, totalMessagesSent,
+                            "group1", consumerNumber, 100, true, maxBytes)));
 
-            int totalResult2 = 0;
-            for (CompletableFuture<Integer> future : futures2) {
-                totalResult2 += future.get();
-            }
+            consumeMessagesFutures2.add(CompletableFuture.supplyAsync(() ->
+                    consumeMessages(totalMessagesConsumedGroup2, totalMessagesSent,
+                            "group2", consumerNumber, 100, true, maxBytes)));
 
-            int totalResult3 = 0;
-            for (CompletableFuture<Integer> future : futures3) {
-                totalResult3 += future.get();
-            }
+            consumeMessagesFutures3.add(CompletableFuture.supplyAsync(() ->
+                    consumeMessages(totalMessagesConsumedGroup3, totalMessagesSent,
+                            "group3", consumerNumber, 100, true, maxBytes)));
+        }
 
-            assertEquals(totalMessagesSent, totalMessagesConsumedGroup1.get());
-            assertEquals(totalMessagesSent, totalMessagesConsumedGroup2.get());
-            assertEquals(totalMessagesSent, totalMessagesConsumedGroup3.get());
-            assertEquals(totalMessagesSent, totalResult1);
-            assertEquals(totalMessagesSent, totalResult2);
-            assertEquals(totalMessagesSent, totalResult3);
+        CompletableFuture.allOf(Stream.of(consumeMessagesFutures1.stream(), consumeMessagesFutures2.stream(),
+                        consumeMessagesFutures3.stream()).flatMap(Function.identity()).toArray(CompletableFuture[]::new))
+                .get(120, TimeUnit.SECONDS);
 
-            int actualMessagesSent = 0;
-            try {
-                producerExecutorService.awaitTermination(60, TimeUnit.SECONDS); // Wait for all producer threads to complete
+        int totalResult1 = consumeMessagesFutures1.stream().map(CompletableFuture::join).reduce(Integer::sum).orElse(0);
+        int totalResult2 = consumeMessagesFutures2.stream().map(CompletableFuture::join).reduce(Integer::sum).orElse(0);
+        int totalResult3 = consumeMessagesFutures3.stream().map(CompletableFuture::join).reduce(Integer::sum).orElse(0);
 
-                for (CompletableFuture<Integer> future : producerFutures) {
-                    actualMessagesSent += future.get();
-                }
-            } catch (Exception e) {
-                fail("Exception occurred : " + e.getMessage());
-            }
-            assertEquals(totalMessagesSent, actualMessagesSent);
-        } catch (Exception e) {
-            fail("Exception occurred : " + e.getMessage());
-        }
+        assertEquals(totalMessagesSent, totalMessagesConsumedGroup1.get());
```

Comment:
Do we actually need `totalMessagesConsumedGroup1`, or can we simply count the returned values from all `consumeMessages` instead?

Reply:
Perhaps we can remove all AtomicInteger here, like you said, now we can rely on the returned values to verify how much records are consumed. WDYT ?

Reply:
I just realized that we cannot remove the AtomicInteger in consumeMessages because it is essential for managing the total number of messages consumed across multiple calls to consumeMessages.

Without it, many tests would hang since the condition `totalMessagesConsumed.get() < totalMessages` would no longer be met, causing the while loop to run indefinitely [0].

[0] https://github.com/apache/kafka/blob/trunk/core/src/test/java/kafka/test/api/ShareConsumerTest.java#L1888

But you are right, the atomic integer assertions seems a bit redundant as now we can rely on the return value from consumeMessages. I'll remove those assertions, thanks.